# Plagiarism Detection Model

Now that you've created training and test data, you are ready to define and train a model. Your goal in this notebook, will be to train a binary classification model that learns to label an answer file as either plagiarized or not, based on the features you provide the model.

This task will be broken down into a few discrete steps:

* Upload your data to S3.
* Define a binary classification model and a training script.
* Train your model and deploy it.
* Evaluate your deployed classifier and answer some questions about your approach.

To complete this notebook, you'll have to complete all given exercises and answer all the questions in this notebook.
> All your tasks will be clearly labeled **EXERCISE** and questions as **QUESTION**.

It will be up to you to explore different classification models and decide on a model that gives you the best performance for this dataset.

---

## Load Data to S3

In the last notebook, you should have created two files: a `training.csv` and `test.csv` file with the features and class labels for the given corpus of plagiarized/non-plagiarized text data. 

>The below cells load in some AWS SageMaker libraries and creates a default bucket. After creating this bucket, you can upload your locally stored data to S3.

Save your train and test `.csv` feature files, locally. To do this you can run the second notebook "2_Plagiarism_Feature_Engineering" in SageMaker or you can manually upload your files to this notebook using the upload icon in Jupyter Lab. Then you can upload local files to S3 by using `sagemaker_session.upload_data` and pointing directly to where the training data is saved.

In [1]:
import pandas as pd
import boto3
import sagemaker

In [2]:
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
# session and role
sagemaker_session = sagemaker.Session()
role = sagemaker.get_execution_role()

# create an S3 bucket
bucket = sagemaker_session.default_bucket()

## EXERCISE: Upload your training data to S3

Specify the `data_dir` where you've saved your `train.csv` file. Decide on a descriptive `prefix` that defines where your data will be uploaded in the default S3 bucket. Finally, create a pointer to your training data by calling `sagemaker_session.upload_data` and passing in the required parameters. It may help to look at the [Session documentation](https://sagemaker.readthedocs.io/en/stable/session.html#sagemaker.session.Session.upload_data) or previous SageMaker code examples.

You are expected to upload your entire directory. Later, the training script will only access the `train.csv` file.

In [3]:
import os

# should be the name of directory you created to save your features data
data_dir = 'plagiarism_data'

# set prefix, a descriptive name for a directory  
prefix = 'plagiarism'

# upload all data to S3

train_input_path = sagemaker_session.upload_data(os.path.join(data_dir,'train.csv'), key_prefix=prefix)
test_input_path = sagemaker_session.upload_data(os.path.join(data_dir,'test.csv'), bucket=bucket, key_prefix=prefix)

print(train_input_path, test_input_path)

s3://sagemaker-us-east-1-903656752510/plagiarism/train.csv s3://sagemaker-us-east-1-903656752510/plagiarism/test.csv


### Test cell

Test that your data has been successfully uploaded. The below cell prints out the items in your S3 bucket and will throw an error if it is empty. You should see the contents of your `data_dir` and perhaps some checkpoints. If you see any other files listed, then you may have some old model files that you can delete via the S3 console (though, additional files shouldn't affect the performance of model developed in this notebook).

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
# confirm that data is in S3 bucket
empty_check = []
for obj in boto3.resource('s3').Bucket(bucket).objects.all():
    empty_check.append(obj.key)
    print(obj.key)

assert len(empty_check) !=0, 'S3 bucket is empty.'
print('Test passed!')

plagiarism/test.csv
plagiarism/train.csv
sagemaker-scikit-learn-2021-04-15-21-05-41-774/source/sourcedir.tar.gz
sagemaker-scikit-learn-2021-04-15-21-22-12-856/source/sourcedir.tar.gz
Test passed!


---

# Modeling

Now that you've uploaded your training data, it's time to define and train a model!

The type of model you create is up to you. For a binary classification task, you can choose to go one of three routes:
* Use a built-in classification algorithm, like LinearLearner.
* Define a custom Scikit-learn classifier, a comparison of models can be found [here](https://scikit-learn.org/stable/auto_examples/classification/plot_classifier_comparison.html).
* Define a custom PyTorch neural network classifier. 

It will be up to you to test out a variety of models and choose the best one. Your project will be graded on the accuracy of your final model. 
 
---



## EXERCISE: Complete a training script 

To implement a custom classifier, you'll need to complete a `train.py` script. You've been given the folders `source_sklearn` and `source_pytorch` which hold starting code for a custom Scikit-learn model and a PyTorch model, respectively. Each directory has a `train.py` training script. To complete this project **you only need to complete one of these scripts**; the script that is responsible for training your final model.

A typical training script:
* Loads training data from a specified directory
* Parses any training & model hyperparameters (ex. nodes in a neural network, training epochs, etc.)
* Instantiates a model of your design, with any specified hyperparams
* Trains that model 
* Finally, saves the model so that it can be hosted/deployed, later

### Defining and training a model
Much of the training script code is provided for you. Almost all of your work will be done in the `if __name__ == '__main__':` section. To complete a `train.py` file, you will:
1. Import any extra libraries you need
2. Define any additional model training hyperparameters using `parser.add_argument`
2. Define a model in the `if __name__ == '__main__':` section
3. Train the model in that same section

Below, you can use `!pygmentize` to display an existing `train.py` file. Read through the code; all of your tasks are marked with `TODO` comments. 

**Note: If you choose to create a custom PyTorch model, you will be responsible for defining the model in the `model.py` file,** and a `predict.py` file is provided. If you choose to use Scikit-learn, you only need a `train.py` file; you may import a classifier from the `sklearn` library.

For this exercise, I have chosen to use a custom Scikit-learn classifier based on Support Vector Machine. Since SVM is a powerful algorithm for classification and it is effective in cases where the number of features is higher. SVM works relatively well when there is a clear margin of separation between classes.

The only problem is that it doesn’t perform well when we have large data set because the required training time is higher, however we have only 70 training set items, that won't be an issue.


In [11]:
# directory can be changed to: source_sklearn or source_pytorch
!pygmentize source_sklearn/train.py

from __future__ import print_function

import argparse
import os
import pandas as pd

# sklearn.externals.joblib is deprecated in 0.21 and will be removed in 0.23. 
# from sklearn.externals import joblib
# Import joblib package directly
import joblib

## TODO: Import any additional libraries you need to define a model
from sklearn.svm import SVC

# Provided model load function
def model_fn(model_dir):
    """Load model from the model_dir. This is the same model that is saved
    in the main if statement.
    """
    print("Loading model.")
    
    # load using joblib
    model = joblib.load(os.path.join(model_dir, "model.joblib"))
    print("Done loading model.")
    
    return model


## TODO: Complete the main code
if __name__ == '__main__':
    
    # All of the model parameters and training parameters are sent as arguments
    # when this script is executed, during a training job
    
    # Here we set up an argument parser to easily access the parameters
    parser = argparse.Ar

### Provided code

If you read the code above, you can see that the starter code includes a few things:
* Model loading (`model_fn`) and saving code
* Getting SageMaker's default hyperparameters
* Loading the training data by name, `train.csv` and extracting the features and labels, `train_x`, and `train_y`

If you'd like to read more about model saving with [joblib for sklearn](https://scikit-learn.org/stable/modules/model_persistence.html) or with [torch.save](https://pytorch.org/tutorials/beginner/saving_loading_models.html), click on the provided links.

---
# Create an Estimator

When a custom model is constructed in SageMaker, an entry point must be specified. This is the Python file which will be executed when the model is trained; the `train.py` function you specified above. To run a custom training script in SageMaker, construct an estimator, and fill in the appropriate constructor arguments:

* **entry_point**: The path to the Python script SageMaker runs for training and prediction.
* **source_dir**: The path to the training script directory `source_sklearn` OR `source_pytorch`.
* **role**: Role ARN, which was specified, above.
* **train_instance_count**: The number of training instances (should be left at 1).
* **train_instance_type**: The type of SageMaker instance for training. Note: Because Scikit-learn does not natively support GPU training, Sagemaker Scikit-learn does not currently support training on GPU instance types.
* **sagemaker_session**: The session used to train on Sagemaker.
* **hyperparameters** (optional): A dictionary `{'name':value, ..}` passed to the train function as hyperparameters.

Note: For a PyTorch model, there is another optional argument **framework_version**, which you can set to the latest version of PyTorch, `1.0`.

## EXERCISE: Define a Scikit-learn or PyTorch estimator

To import your desired estimator, use one of the following lines:
```
from sagemaker.sklearn.estimator import SKLearn
```
```
from sagemaker.pytorch import PyTorch
```

In [12]:
# your import and estimator code, here

from sagemaker.sklearn.estimator import SKLearn


sklearn_estimator = SKLearn(entry_point='train.py',
                            source_dir='source_sklearn',
                            framework_version='0.23-1',
                            role=role,
                            train_instance_count=1,
                            train_instance_type='ml.m4.xlarge',
                            sagemaker_session=sagemaker_session,
                            hyperparameters={
                                'c': 1.0,
                                'kernel' : 'rbf'
                            })

train_instance_type has been renamed in sagemaker>=2.
See: https://sagemaker.readthedocs.io/en/stable/v2.html for details.
train_instance_count has been renamed in sagemaker>=2.
See: https://sagemaker.readthedocs.io/en/stable/v2.html for details.
train_instance_count has been renamed in sagemaker>=2.
See: https://sagemaker.readthedocs.io/en/stable/v2.html for details.
train_instance_type has been renamed in sagemaker>=2.
See: https://sagemaker.readthedocs.io/en/stable/v2.html for details.


## EXERCISE: Train the estimator

Train your estimator on the training data stored in S3. This should create a training job that you can monitor in your SageMaker console.

In [14]:
%%time

# Train your estimator on S3 training data

sklearn_estimator.fit({'train': train_input_path})

2021-04-16 16:24:42 Starting - Starting the training job...
2021-04-16 16:24:53 Starting - Launching requested ML instancesProfilerReport-1618590281: InProgress
......
2021-04-16 16:26:07 Starting - Preparing the instances for training.........
2021-04-16 16:27:39 Downloading - Downloading input data...
2021-04-16 16:28:09 Training - Downloading the training image...
2021-04-16 16:28:41 Uploading - Uploading generated training model2021-04-16 16:28:31,570 sagemaker-containers INFO     Imported framework sagemaker_sklearn_container.training
2021-04-16 16:28:31,573 sagemaker-training-toolkit INFO     No GPUs detected (normal if no gpus installed)
2021-04-16 16:28:31,584 sagemaker_sklearn_container.training INFO     Invoking user training script.
2021-04-16 16:28:31,935 sagemaker-training-toolkit INFO     No GPUs detected (normal if no gpus installed)
2021-04-16 16:28:34,968 sagemaker-training-toolkit INFO     No GPUs detected (normal if no gpus installed)
2021-04-16 16:28:34,982 sagemake

## EXERCISE: Deploy the trained model

After training, deploy your model to create a `predictor`. If you're using a PyTorch model, you'll need to create a trained `PyTorchModel` that accepts the trained `<model>.model_data` as an input parameter and points to the provided `source_pytorch/predict.py` file as an entry point. 

To deploy a trained model, you'll use `<model>.deploy`, which takes in two arguments:
* **initial_instance_count**: The number of deployed instances (1).
* **instance_type**: The type of SageMaker instance for deployment.

Note: If you run into an instance error, it may be because you chose the wrong training or deployment instance_type. It may help to refer to your previous exercise code to see which types of instances we used.

In [15]:
%%time

# deploy your model to create a predictor

predictor = sklearn_estimator.deploy(initial_instance_count=1, instance_type='ml.t2.medium')


---------------------!CPU times: user 362 ms, sys: 11.5 ms, total: 373 ms
Wall time: 10min 32s


---
# Evaluating Your Model

Once your model is deployed, you can see how it performs when applied to our test data.

The provided cell below, reads in the test data, assuming it is stored locally in `data_dir` and named `test.csv`. The labels and features are extracted from the `.csv` file.

In [16]:
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import os

# read in test data, assuming it is stored locally
test_data = pd.read_csv(os.path.join(data_dir, "test.csv"), header=None, names=None)

# labels are in the first column
test_y = test_data.iloc[:,0]
test_x = test_data.iloc[:,1:]

## EXERCISE: Determine the accuracy of your model

Use your deployed `predictor` to generate predicted, class labels for the test data. Compare those to the *true* labels, `test_y`, and calculate the accuracy as a value between 0 and 1.0 that indicates the fraction of test data that your model classified correctly. You may use [sklearn.metrics](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.metrics) for this calculation.

**To pass this project, your model should get at least 90% test accuracy.**

In [18]:
# First: generate predicted, class labels

test_y_preds = predictor.predict(test_x)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
# test that your model generates the correct number of labels
assert len(test_y_preds)==len(test_y), 'Unexpected number of predictions.'
print('Test passed!')

Test passed!


In [19]:
# Second: calculate the test accuracy
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(test_y_preds, test_y)

print(accuracy)

## print out the array of predicted and true labels, if you want

print('\nPredicted class labels: ')
print(test_y_preds)
print('\nTrue class labels: ')
print(test_y.values)

0.96

Predicted class labels: 
[1 1 1 1 1 1 0 0 0 0 0 0 0 1 1 1 1 1 0 1 0 1 1 0 0]

True class labels: 
[1 1 1 1 1 1 0 0 0 0 0 0 1 1 1 1 1 1 0 1 0 1 1 0 0]


In [22]:
# code to evaluate the endpoint on test data
# returns a variety of model metrics

import numpy as np

def evaluate(predictor, test_features, test_labels, verbose=True):
    """
    Evaluate a model on a test set given the prediction endpoint.  
    Return binary classification metrics.
    :param predictor: A prediction endpoint
    :param test_features: Test features
    :param test_labels: Class labels for test data
    :param verbose: If True, prints a table of all performance metrics
    :return: A dictionary of performance metrics.
    """
    
    # rounding and squeezing array
    test_preds = np.squeeze(np.round(predictor.predict(test_features)))
    
    # calculate true positives, false positives, true negatives, false negatives
    tp = np.logical_and(test_labels, test_preds).sum()
    fp = np.logical_and(1-test_labels, test_preds).sum()
    tn = np.logical_and(1-test_labels, 1-test_preds).sum()
    fn = np.logical_and(test_labels, 1-test_preds).sum()
    
    # calculate binary classification metrics
    recall = tp / (tp + fn)
    precision = tp / (tp + fp)
    accuracy = (tp + tn) / (tp + fp + tn + fn)
    
    # print metrics
    if verbose:
        print(pd.crosstab(test_labels, test_preds, rownames=['actuals'], colnames=['predictions']))
        print("\n{:<11} {:.3f}".format('Recall:', recall))
        print("{:<11} {:.3f}".format('Precision:', precision))
        print("{:<11} {:.3f}".format('Accuracy:', accuracy))
        print()
        
    return {'TP': tp, 'FP': fp, 'FN': fn, 'TN': tn, 
            'Precision': precision, 'Recall': recall, 'Accuracy': accuracy}


In [23]:
evaluate(predictor,test_x,test_y, verbose=True)

predictions   0   1
actuals            
0            10   0
1             1  14

Recall:     0.933
Precision:  1.000
Accuracy:   0.960



{'TP': 14,
 'FP': 0,
 'FN': 1,
 'TN': 10,
 'Precision': 1.0,
 'Recall': 0.9333333333333333,
 'Accuracy': 0.96}

In [30]:
results = pd.concat([test_x, test_y, pd.DataFrame(test_y_preds)], axis=1)
results.columns = ['c_1','c_14','c_19', 'lcs','class','prediction']
results

c_1      c_14      c_19       lcs  class  prediction
0   1.000000  0.722826  0.603352  0.820755      1           1
1   0.765306  0.348754  0.286232  0.621711      1           1
2   0.884444  0.000000  0.000000  0.597458      1           1
3   0.619048  0.000000  0.000000  0.427835      1           1
4   0.920000  0.064516  0.000000  0.775000      1           1
5   0.992674  0.938462  0.917647  0.993056      1           1
6   0.412698  0.000000  0.000000  0.346667      0           0
7   0.462687  0.000000  0.000000  0.189320      0           0
8   0.581152  0.000000  0.000000  0.247423      0           0
9   0.584211  0.000000  0.000000  0.294416      0           0
10  0.566372  0.000000  0.000000  0.258333      0           0
11  0.481481  0.000000  0.000000  0.278912      0           0
12  0.619792  0.000000  0.000000  0.341584      1           0
13  0.921739  0.294931  0.179245  0.929412      1           1
14  1.000000  0.738956  0.643443  1.000000      1           1
15  0.861538  0.000000  0.000000  0.504717      1           1
16  0.626168  0.006494  0.000000  0.558559      1           1
17  1.000000  0.895161  0.851852  0.996700      1           1
18  0.383838  0.000000  0.000000  0.178744      0           0
19  1.000000  0.805344  0.731518  0.854671      1           1
20  0.613924  0.000000  0.000000  0.298343      0           0
21  0.972763  0.557377  0.468619  0.927083      1           1
22  0.962810  0.349345  0.263393  0.909804      1           1
23  0.415254  0.000000  0.000000  0.177419      0           0
24  0.532189  0.000000  0.000000  0.245833      0           0

### Question 1: How many false positives and false negatives did your model produce, if any? And why do you think this is?

** Answer**: 

Out of 25 items in test data, 24 are predicted correctly, hence the model get 96% of accuracy with only 1 text is classified incorrectly - false negative.

True Positives: 14,
False Positives: 0,
False Negatives: 1,
True Negatives: 10

While comparing the features of all the test data, the values for the misclassified text (row 12 predicted as non-plagiarised) are low compared to the values of other plagiarised text:

c_1 = 0.619792, c_14 = 0.000000, c_19 = 0.000000, lcs = 0.341584

And the misclassified answer is classified as heavy modified by changing the words.

### Question 2: How did you decide on the type of model to use? 

** Answer**:

As I explained above while deciding on the type of algorithm, I have chosen to use a custom Scikit-learn classifier based on Support Vector Machine. Since SVM is a powerful algorithm for classification and it is effective in cases where the number of features is higher. SVM works relatively well when there is a clear margin of separation between classes.

Also SVM fits well on small data sets. The data sets for this problem is very small (70 training set items and 25 test set items) which makes it a good fit.

----
## EXERCISE: Clean up Resources

After you're done evaluating your model, **delete your model endpoint**. You can do this with a call to `.delete_endpoint()`. You need to show, in this notebook, that the endpoint was deleted. Any other resources, you may delete from the AWS console, and you will find more instructions on cleaning up all your resources, below.

In [31]:
# uncomment and fill in the line below!
# <name_of_deployed_predictor>.delete_endpoint()

predictor.delete_endpoint()


### Deleting S3 bucket

When you are *completely* done with training and testing models, you can also delete your entire S3 bucket. If you do this before you are done training your model, you'll have to recreate your S3 bucket and upload your training data again.

In [32]:
# deleting bucket, uncomment lines below

bucket_to_delete = boto3.resource('s3').Bucket(bucket)
bucket_to_delete.objects.all().delete()

[{'ResponseMetadata': {'RequestId': 'F533V3WS6CWVVY8K',
   'HostId': 'VvMEA4fgM+NQ1BazNoEVM5e7/PGWHyiJYaNP9BcumNbbWd+1DKz89RRDQ/coLva/7B2T4g8JBoQ=',
   'HTTPStatusCode': 200,
   'HTTPHeaders': {'x-amz-id-2': 'VvMEA4fgM+NQ1BazNoEVM5e7/PGWHyiJYaNP9BcumNbbWd+1DKz89RRDQ/coLva/7B2T4g8JBoQ=',
    'x-amz-request-id': 'F533V3WS6CWVVY8K',
    'date': 'Fri, 16 Apr 2021 17:14:19 GMT',
    'content-type': 'application/xml',
    'transfer-encoding': 'chunked',
    'server': 'AmazonS3',
    'connection': 'close'},
   'RetryAttempts': 0},
  'Deleted': [{'Key': 'sagemaker-scikit-learn-2021-04-15-21-05-41-774/source/sourcedir.tar.gz'},
   {'Key': 'sagemaker-scikit-learn-2021-04-16-15-51-03-551/rule-output/ProfilerReport-1618588263/profiler-output/profiler-reports/LowGPUUtilization.json'},
   {'Key': 'sagemaker-scikit-learn-2021-04-16-15-51-03-551/rule-output/ProfilerReport-1618588263/profiler-output/profiler-reports/Dataloader.json'},
   {'Key': 'sagemaker-scikit-learn-2021-04-16-16-24-41-873/profiler-

### Deleting all your models and instances

When you are _completely_ done with this project and do **not** ever want to revisit this notebook, you can choose to delete all of your SageMaker notebook instances and models by following [these instructions](https://docs.aws.amazon.com/sagemaker/latest/dg/ex1-cleanup.html). Before you delete this notebook instance, I recommend at least downloading a copy and saving it, locally.

---
## Further Directions

There are many ways to improve or add on to this project to expand your learning or make this more of a unique project for you. A few ideas are listed below:
* Train a classifier to predict the *category* (1-3) of plagiarism and not just plagiarized (1) or not (0).
* Utilize a different and larger dataset to see if this model can be extended to other types of plagiarism.
* Use language or character-level analysis to find different (and more) similarity features.
* Write a complete pipeline function that accepts a source text and submitted text file, and classifies the submitted text as plagiarized or not.
* Use API Gateway and a lambda function to deploy your model to a web application.

These are all just options for extending your work. If you've completed all the exercises in this notebook, you've completed a real-world application, and can proceed to submit your project. Great job!